libkbfs/test_stallers.go

```diff
@@ -38,6 +39,166 @@ const (
+// StallBlockOp and StallMDOp which provides a way to precisely control which
+// particular op is stalled by passing in ctx with corresponding stallKey,
+// NaïveStaller simply stalls all instances of specified op.
+type NaïveStaller struct {
+	config Config
+
+	mu             *sync.RWMutex
+	blockOpsStalls map[StallableBlockOp]*naïveStallContext
+	mdOpsStalls    map[StallableMDOp]*naïveStallContext
+}
+
+// NewNaïveStaller returns a new NaïveStaller
+func NewNaïveStaller(config Config) *NaïveStaller {
+	return &NaïveStaller{
+		config:         config,
+		mu:             new(sync.RWMutex),
+		blockOpsStalls: make(map[StallableBlockOp]*naïveStallContext),
+		mdOpsStalls:    make(map[StallableMDOp]*naïveStallContext),
+	}
+}
+
+func (s *NaïveStaller) getNaïveStallerContextForBlockOp(
+	stalledOp StallableBlockOp) *naïveStallContext {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.blockOpsStalls[stalledOp]
+}
+
+func (s *NaïveStaller) getNaïveStallerContextForMDOp(
+	stalledOp StallableMDOp) *naïveStallContext {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.mdOpsStalls[stalledOp]
+}
+
+// StallBlockOp wraps the internal BlockOps so that all subsequent stalledOp
+// will be stalled. This can be undone by calling UndoStallBlockOp.
+func (s *NaïveStaller) StallBlockOp(stalledOp StallableBlockOp) {
+	onStalledCh := make(chan struct{}, 1)
+	unstallCh := make(chan struct{})
+	oldBlockOps := s.config.BlockOps()
+	s.config.SetBlockOps(&stallingBlockOps{
+		stallOpName: stalledOp,
+		stallKey:    "",
+		staller: staller{
+			stalled: onStalledCh,
+			unstall: unstallCh,
+		},
+		delegate: oldBlockOps,
+	})
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.blockOpsStalls[stalledOp] = &naïveStallContext{
+		onStalled: onStalledCh,
+		unstall:   unstallCh,
+		undoStall: func() {
+			s.config.SetBlockOps(oldBlockOps)
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			delete(s.blockOpsStalls, stalledOp)
+		},
+	}
+}
+
+// StallMDOp wraps the internal MDOps so that all subsequent stalledOp
+// will be stalled. This can be undone by calling UndoStallMDOp.
+func (s *NaïveStaller) StallMDOp(stalledOp StallableMDOp) {
+	onStalledCh := make(chan struct{}, 1)
+	unstallCh := make(chan struct{})
+	oldMDOps := s.config.MDOps()
+	s.config.SetMDOps(&stallingMDOps{
+		stallOpName: stalledOp,
+		stallKey:    "",
+		staller: staller{
+			stalled: onStalledCh,
+			unstall: unstallCh,
+		},
+		delegate: oldMDOps,
+	})
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.mdOpsStalls[stalledOp] = &naïveStallContext{
+		onStalled: onStalledCh,
+		unstall:   unstallCh,
+		undoStall: func() {
+			s.config.SetMDOps(oldMDOps)
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			delete(s.mdOpsStalls, stalledOp)
+		},
+	}
+}
+
+// WaitForStallBlockOp blocks until stalledOp is stalled. StallBlockOp should
+// have been called upon stalledOp, otherwise this is a no-op.
+func (s *NaïveStaller) WaitForStallBlockOp(stalledOp StallableBlockOp) {
+	ctx := s.getNaïveStallerContextForBlockOp(stalledOp)
+	if ctx == nil {
```

Comment:
I wonder if it would be better to simply panic here, or in `getNaiveStaller...`? It seems that anyone who calls this function expects `StallBlockOp` to have been called, and no-oping would just simply cause a hang further down the line. (Same everywhere else.)